chronos/streams/asyncstream.nim

```diff
@@ -141,12 +141,37 @@ proc copyData*(sb: AsyncBuffer, dest: pointer, offset, length: int) {.inline.} =
   copyMem(cast[pointer](cast[uint](dest) + cast[uint](offset)),
           unsafeAddr sb.buffer[0], length)
 
+proc upload*(sb: ptr AsyncBuffer, pbytes: ptr byte,
+             nbytes: int): Future[void] {.async.} =
+  var length = nbytes
+  while length > 0:
+    let size = min(length, sb[].bufferLen())
+    if size == 0:
+      # Internal buffer is full, we need to transfer data to consumer.
+      await sb[].transfer()
+      continue
+    else:
+      copyMem(addr sb[].buffer[sb.offset], pbytes, size)
+      sb[].offset = sb[].offset + size
+      length = length - size
+  # We notify consumers that new data is available.
+  sb[].forget()
+
 template toDataOpenArray*(sb: AsyncBuffer): auto =
   toOpenArray(sb.buffer, 0, sb.offset - 1)
 
 template toBufferOpenArray*(sb: AsyncBuffer): auto =
   toOpenArray(sb.buffer, sb.offset, len(sb.buffer) - 1)
 
+template harvestItem*(dest: pointer, item: WriteItem, length: int) =
```

Comment:
I think `copyOut` is a more commonly accepted name for the operation (`item` arg goes in front). Also please avoid templates in favor of procs (optionally inline).

Reply:
Every copy procedure uses destination as first item, `item` here is source.